daemons/gptp/common/ieee1588clock.cpp

```diff
@@ -89,6 +91,21 @@ IEEE1588Clock::IEEE1588Clock
 
 	this->forceOrdinarySlave = forceOrdinarySlave;
 
+    if (externalPortConfiguration == EXT_DISABLED) {
```

Comment:
In the current draft of 802.1AS-revision, externalPortConfiguration is a feature that is analogous to the "disable BMCA" feature of the Avnu Automotive Profile. Therefore, externalPortConfiguration of ENABLED turns off the BMCA, but it does nothing to other features like transmitAnnounce, forceAsCapable, and so on, so this conditional is incorrect.

To backup a bit and discuss the high-level concepts, we have:
- "Feature": An optional feature of the standard, that specifies how the feature works, what is required to implement the feature, and so on.
- "Profile": A specification that says "for my application (e.g. inside a car), optional features X, Y, and Z shall be configured this way". The profile does not add new features... it merely says how the features are used.

The Anvu Automotive Profile did both: it added new features to 802.1AS, and it specified a profile. That's okay, but we need to enhance the code, because:
- One of its features is now part of the 802.1AS standard (i.e. externalPortConfiguration).
- Some applications want to use some of the new Anvu Automotive Profile features, but not all.
The first step is to provide a variable to configure each feature separately, which you have.

The second step is to retain the "profile". I recommend the following changes:
- Add the "automotive_profile" boolean back. This represents the Avnu Automotive Profile.
- Change this conditional to "if (automotive_profile) {"
    This makes it a true profile... it just specifies what features are used and how.
- Add the following line inside the conditional: "externalPortConfiguration = ENABLED;"
    The Avnu Automotive Profile disables the BMCA, and this accomplishes that requirement.

Reply:
Hi Rodney,
Thank you for clarifying the concepts. I have added back the "automotive_profile" boolean to separate the extPortConfig feature and the automotive profile. The changes are included in the latest commit.
I have a question about one specific use case. In our implementation, the ExtPortConfig Enum has 3 states: DISABLED, GM (extPortConfig is enabled and port state is forced to be master), SLAVE( extPortConfig is enabled and port state is forced to be slave). It is designed to reduce the number of variables for IEEE1588Clock initialization. If automotive_profile is enabled while extPortConfig is disabled by default, what port_state should we assign to  extPortConfig when we enable it？
In the latest commit, I just added a new state in the ExtPortConfig Enum "ENABLED (extPortConfig is enabled but the port state is not specified)" and port state is set to PTP_INITIALIZING in this case.
What's your suggestion on this problem?

Reply:
Hi Yaqi,
If possible, I think it would be best to align the variables with the names from the standards. That way, when someone reads the standard it matches the code. Both IEEE 1588-rev and IEEE 802.1AS-rev will have "externalPortConfiguration" as an enumeration, with values ENABLED and DISABLED (only). Both also have portState with MASTER, SLAVE, and so on.

You can take a look in Perforce at what I did for a development branch for a customer:
  //ThirdPartyExports/gPTP/development/16.1/cummings_external_port/...
I think I passed portState and externalPortConfiguration in from the constructor, so I always had values from the user. In your case you would pass both in from the command-line.